evennia/contrib/pathfinding/pathfinder.py

```diff
@@ -0,0 +1,394 @@
+        self.key = attrs.get('key', attrs.get('direction'))
+        self.lock_storage = attrs['locks_str']
+        self.locks = LockHandler(self)
+
+class Pathfinder(nx.DiGraph):
+    """
+    Creates a directional graph of your entire network of rooms and exits.
+    """
+    # Arbitrary delimiter used to create unique keys to use as nodes.
+    # Should not be set it to anything that could conceivably be part of a room
+    # name since it will interfere with parsing.
+    delimiter = '[(:@_@:)]'
+    
+    def __init__(self, *args, **kwargs):
+        """
+        Creates the graph and initializes it.
+        
+        Kwargs:
+            queryset (QuerySet, optional): QuerySet of Room objects to be 
+                included in the graph
+                
+        """
+        self.queryset = kwargs.pop('queryset', None)
+        super(Pathfinder, self).__init__(*args, **kwargs)
+        
+    def get_key(self, room):
+        """
+        Creates unique keys for use as nodes in the graph.
+
+        Args:
+            room (Room): Room object needing a key.
+            
+        Returns:
+            key (str): Unique key identifying the node.
+        
+        """
+        return self.delimiter.join([room.db_key, str(room.id)])
+        
+    def get_queryset(self):
+        """
+        Get all the Room objects to be included in this graph.
+        
+        Returns:
+            objs (generator): Any iterable generator of Room objects you
+                wish to be part of the graph.
+            
+        """
+        if self.queryset: 
+            return (x for x in self.queryset)
+        else:
+            return Room.objects.all().iterator()
+    
+    def update(self):
+        """
+        Updates the graph, adding or deleting any new/old Rooms and Exits.
+
+        Returns:
+            self (Pathfinder): This object.
+
+        """
+        # Get all Room objects in play
+        all_rooms = self.get_queryset()
+        
+        for src in all_rooms:
+            # Create node for src room
+            src_key = self.get_key(src)
+            
+            # Check if node exists
+            try: self[src_key]
+            except: self.add_node(src_key, key=src.db_key, id=src.id, type='room', locks_str=str(src.locks))
+            
+            for exit in src.exits:
+                # Get destination room
+                dst = exit.destination
+                
+                # Create a node for dst room
+                dst_key = self.get_key(dst)
+                try: self[dst_key]
+                except: self.add_node(dst_key, key=dst.db_key, id=dst.id, type='room', locks_str=str(dst.locks))
+                
+                # Create an edge representing the exit
+                self.add_edge(src_key, dst_key, direction=exit.db_key, id=exit.id, type='exit', locks_str=str(exit.locks))
+                
+        return self
+                
+    def get_path(self, source, dest):
```

Comment:
Similarly to the line-of-sight calculation, how does this handle cases where a room (say, on the way to the destination), has exits that loop back on itself or to a room earlier in the path?

Reply:
I understand your question but I'm having a hard time simulating it, so I suspect networkx is refusing to chase its own tail in this manner. An infinite loop is never a valid path, and I can't seem to make it get stuck trying to traverse one.

LOS calculation was a bit different as that was my own implementation and I hadn't thought to do such checks when you raised that issue, but in other contexts nx is smart enough to return "inf" (infinity) as the length of a path so given my inability to break it I'm assuming it's doing the same here.

Reply:
Ok, good to know! Worthy of a note in the docstring to show that it was considered in case people wonder later (and they will). :)